test/com/amazon/ion/impl/IonCursorBinaryTest.java

```diff
@@ -97,15 +117,11 @@ public void basicStrings() {
             0x83, 'f', 'o', 'o', // String length 3, starting at byte index 5
             0x83, 'b', 'a', 'r' // String length 3, starting at byte index 9
         );
-        assertEquals(START_SCALAR, cursor.nextValue());
-        Marker marker = loadScalar(cursor);
-        assertEquals(5, marker.startIndex);
-        assertEquals(8, marker.endIndex);
-        assertEquals(START_SCALAR, cursor.nextValue());
-        marker = loadScalar(cursor);
-        assertEquals(9, marker.startIndex);
-        assertEquals(12, marker.endIndex);
-        assertEquals(NEEDS_DATA, cursor.nextValue());
+        nextExpect(START_SCALAR);
+        expectMarker(5, 8);
+        nextExpect(START_SCALAR);
+        expectMarker(9, 12);
+        nextExpect(NEEDS_DATA);
```

Comment:
I'd rather see something like this:
```java
assertSequence(cursor,
  marker(5,8),
  marker(9,12),
  end()
)
```

Difference being that the cursor is no longer implicit, and instead the assertion helper methods are creating a sequence of cursor consumers. It leads nicely into a succinct expression of what you expect to happen. 

```java
public void assertSequence(IonCursor cursor, Consumer<IonCursor>... expectations) {
  expectations.forEach( e -> e.accept(cursor));
}

public Consumer<IonCursor> compose(Consumer<IonCursor>... expectations) {
  return c -> expectations.forEach( e -> e.accept(cursor));
}

public Consumer<IonCursor> marker(int expectedStart, int expectedEnd) {
  return compose(START_SCALAR,
    c -> {
      Marker marker = loadScalar(cursor);
      assertEquals(expectedStart, marker.startIndex);
      assertEquals(expectedEnd, marker.endIndex);
    }
  );
}

public Consumer<IonCursor> container(Consumer<IonCursor>... expectations) {
  return compose(
    START_CONTAINER,
    compose(expectations),
    END_CONTAINER
  );
}

// etc.
```

Then it's also easy to compose consumers, such that you can easily produce assertions like:

```java
assertSequence(cursor,
  container(
    marker(7,8)
  )
)
```

Reply:
The above is assuming reasonable implementations for constant consumer/assertions such as `START_CONTAINER` etc.

Reply:
Here's what I tried:

```java
    @SafeVarargs
    private final void assertSequence(IonCursorBinary cursor, Consumer<IonCursorBinary>... expectations) {
        for (Consumer<IonCursorBinary> expectation : expectations) {
            expectation.accept(cursor);
        }
    }

    @SafeVarargs
    private final Consumer<IonCursorBinary> compose(Consumer<IonCursorBinary>... expectations) {
        return cursor -> assertSequence(cursor, expectations);
    }

    private static final Consumer<IonCursorBinary> SCALAR = c -> assertEquals(START_SCALAR, c.nextValue());
    private static final Consumer<IonCursorBinary> CONTAINER_START = c -> assertEquals(START_CONTAINER, c.nextValue());
    private static final Consumer<IonCursorBinary> CONTAINER_END = c -> assertEquals(END_CONTAINER, c.nextValue());
    private static final Consumer<IonCursorBinary> STREAM_END = c -> assertEquals(NEEDS_DATA, c.nextValue());

    public Consumer<IonCursorBinary> scalarMarker(int expectedStart, int expectedEnd) {
        return compose(
            SCALAR,
            c -> {
                Marker marker = loadScalar(c);
                assertEquals(expectedStart, marker.startIndex);
                assertEquals(expectedEnd, marker.endIndex);
            }
        );
    }

    @SafeVarargs
    private final Consumer<IonCursorBinary> container(Consumer<IonCursorBinary>... expectations) {
        return compose(
            CONTAINER_START,
            IonCursor::stepIntoContainer,
            compose(expectations),
            CONTAINER_END,
            IonCursor::stepOutOfContainer
        );
    }

    @SafeVarargs
    private final void assertStream(IonCursorBinary cursor, Consumer<IonCursorBinary>... expectations) {
        assertSequence(cursor, compose(expectations), STREAM_END);
    }

    @Test
    public void basicContainer() {
        initializeCursor(
            0xE0, 0x01, 0x00, 0xEA,
            0xD3, // Struct length 3
            0x84, // Field SID 4
            0x21, 0x01 // Int length 1, starting at byte index 7
        );
        assertStream(
            cursor,
            container(
                scalarMarker(7, 8)
            )
        );
        /*
        nextExpect(START_CONTAINER);
        stepIn();
        nextExpect(START_SCALAR);
        expectMarker(7, 8);
        nextExpect(END_CONTAINER);
        stepOut();
        nextExpect(NEEDS_DATA);
         */
    }

    @Test
    public void basicStrings() {
        initializeCursor(
            0xE0, 0x01, 0x00, 0xEA,
            0x83, 'f', 'o', 'o', // String length 3, starting at byte index 5
            0x83, 'b', 'a', 'r' // String length 3, starting at byte index 9
        );
        assertStream(
            cursor,
            scalarMarker(5, 8),
            scalarMarker(9, 12)
        );
        /*
        nextExpect(START_SCALAR);
        expectMarker(5, 8);
        nextExpect(START_SCALAR);
        expectMarker(9, 12);
        nextExpect(NEEDS_DATA);
         */
    }
```

This makes the individual tests a bit cleaner, but I have some real concerns about this technique hurting debuggability. For example, let's say my `basicStrings` test starts failing on the second scalar marker. Where do I set my breakpoint? In the old code, I'd set it at `expectMarker(9, 12)`, giving me a direct line to the code under test.

Under this proposal, my instinct is to set it on `assertStream`, but when I step into that all I have to work with is an opaque sequence of Consumers. I have to loop through them until I get to the one I want to debug. I could set breakpoints in the IonCursorBinary methods under test, but these will activate on every invocation both within this test and others; enabling/disabling breakpoints and/or using conditional breakpoints is extra work that adds to debugging overhead.

Am I missing something obvious here? Overall I think this is a cool technique, but I want to avoid accepting a marginal cleanliness/readability improvement at the expense of a major debuggability regression.

Reply:
It _is_ a cool technique. Part of what spurned my initial comment that led to this PR was the concise elegance of the tests in `ion-python`. But you know what? Debugging those tests sucks. For exactly the reason you state here.

If I'm debugging Kotlin in IDEA and using lambdas it's smart enough to let me set a breakpoint when a certain lambda is constructed _or_ invoked (IIRC). If one can't do that with vanilla java in IDE __ then yea that's a reason to prefer the straightforward OO imperative pattern.

Reply:
Fair enough, you definitely gave it a shot. Is debugging usability improved by removing variadic `compose`? Then you'd have a sequence of assertion statements in the body instead of a series of consumers, while still retaining structural associations (e.g. pairing `stepIn()` and `stepOut()`).

If nothing else, it would be good to avoid the implicit cursor pattern in my opinion- passing parameters explicitly really helps with test maintenance.

Reply:
Okay, I have a factoring that mostly addresses my debuggability concerns, at the expense of increased setup code. View it here: https://gist.github.com/tgregg/ba7a908c8407e99201097bce747bf152 

The keys here are the custom Expectation class that provides a helpful toString() implementation, and the collection of all Expectations in a flat list before evaluation. This allows developers to set a breakpoint on `assertSequence` in the test they want to debug, and do a minimal amount of step-in/over to get to the code under test.

I'm still on the fence about whether it's worth the effort of the refactor, but I would like opinions from @jobarr-amzn and @rmarrowstone about whether it's better or worse than what we currently have.

Reply:
I left a couple of comments on the gist. I think that we can improve the debuggability more, and from my point of view it's worth it because it:

1. Creates stronger assurances and removes foot-guns (pairing states where appropriate, etc.)
2. Allows re-use in other contexts
3. Makes parameters explicit rather than implicit, which makes the tests more maintainable

Reply:
I suspect the upcoming work to support Ion 1.1 will pay down any investments in making the test easier to maintain and extend. I personally like the how the expectations (with `container(...`) mirror the expected structure. That makes it easy for me to look at the test code and "see" the expected structure more easily.